src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,158 @@
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventSourceCreated(EventSource source)
+    {
+        if (this.options == null)
+        {
+            this.preInitEventSources.Enqueue(source);
+        }
+        else if (this.options.ShouldListenToSource(source.Name))
+        {
+            this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        if (this.options == null)
+        {
+            return;
+        }
+
+        var eventSourceName = eventData.EventSource.Name;
+
+        if (eventData.EventName != "EventCounters")
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreNonEventCountersName(eventSourceName);
+            return;
+        }
+
+        if (eventData.Payload == null || eventData.Payload.Count == 0 || eventData.Payload[0] is not IDictionary<string, object>)
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenEventArgsPayloadNotParseable(eventSourceName);
+            return;
+        }
+
+        var payload = eventData.Payload[0] as IDictionary<string, object>;
+        var hasName = payload.TryGetValue("Name", out var nameObj);
+
+        if (!hasName)
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenEventArgsWithoutName(eventSourceName);
+            return;
+        }
+
+        var name = nameObj.ToString();
+
+        var hasMean = payload.TryGetValue("Mean", out var mean);
+        var hasIncrement = payload.TryGetValue("Increment", out var increment);
+
+        if (!(hasIncrement ^ hasMean))
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreMeanIncrementConflict(eventSourceName);
+            return;
+        }
+
+        var value = Convert.ToDouble(hasMean ? mean : increment);
+        this.UpdateInstrumentWithEvent(hasMean, eventSourceName, name, value);
+    }
+
+    private static Dictionary<string, string> GetEnableEventsArguments(EventCountersInstrumentationOptions options) =>
+        new() { { "EventCounterIntervalSec", options.RefreshIntervalSecs.ToString() } };
+
+    private static bool IsValidInstrumentName(string name) =>
+        !string.IsNullOrWhiteSpace(name) && InstrumentNameRegex.IsMatch(name);
+
+    private void UpdateInstrumentWithEvent(bool isGauge, string eventSourceName, string name, double value)
+    {
+        try
+        {
+            ValueTuple<string, string> metricKey = new(eventSourceName, name);
+            _ = this.values.AddOrUpdate(metricKey, value, isGauge ? (_, _) => value : (_, existing) => existing + value);
+
+            var instrumentName = $"EventCounters.{eventSourceName}.{name}";
+
+            if (!IsValidInstrumentName(instrumentName))
```

Comment:
You can remove the checks for instrument name. It's already done at the SDK: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MeterProviderSdk.cs#L273-L282